Irrelevant `Cannot compile modules using option 'outFile'` error when using `emitDeclarationOnly`

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

Noticed via https://stackoverflow.com/q/52326378 .

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  master (e471856)

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** cannot compile modules outFile emitDeclarationOnly

**Code**
```ts
// tsconfig.json
{
  "compilerOptions": {
    "declaration": true,
    "emitDeclarationOnly": true,
    "outFile": "out.d.ts"
  }
}
// a.ts
export default 42;
```

**Expected behavior:** Successful declaration file generation.

**Actual behavior:** `a.ts:1:1 - error TS6131: Cannot compile modules using option 'outFile' unless the '--module' flag is 'amd' or 'system'.`

**Playground Link:** N/A

**Related Issues:** None found

--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -2577,7 +2577,7 @@ namespace ts {
             }
 
             // Cannot specify module gen that isn't amd or system with --out
-            if (outFile) {
+            if (outFile && !options.emitDeclarationOnly) {
                 if (options.module && !(options.module === ModuleKind.AMD || options.module === ModuleKind.System)) {
                     createDiagnosticForOptionName(Diagnostics.Only_amd_and_system_modules_are_supported_alongside_0, options.out ? "out" : "outFile", "module");
                 }

--- a/src/compiler/utilities.ts
+++ b/src/compiler/utilities.ts
@@ -3265,7 +3265,7 @@ namespace ts {
         const isSourceFileFromExternalLibrary = (file: SourceFile) => host.isSourceFileFromExternalLibrary(file);
         if (options.outFile || options.out) {
             const moduleKind = getEmitModuleKind(options);
-            const moduleEmitEnabled = moduleKind === ModuleKind.AMD || moduleKind === ModuleKind.System;
+            const moduleEmitEnabled = options.emitDeclarationOnly || moduleKind === ModuleKind.AMD || moduleKind === ModuleKind.System;
             // Can emit only sources that are not declaration file and are either non module code or module with --module or --target es6 specified
             return filter(host.getSourceFiles(), sourceFile =>
                 (moduleEmitEnabled || !isExternalModule(sourceFile)) && sourceFileMayBeEmitted(sourceFile, options, isSourceFileFromExternalLibrary));

--- a/tests/baselines/reference/outModuleConcatCommonjsDeclarationOnly.js
+++ b/tests/baselines/reference/outModuleConcatCommonjsDeclarationOnly.js
@@ -1,0 +1,21 @@
+//// [tests/cases/compiler/outModuleConcatCommonjsDeclarationOnly.ts] ////
+
+//// [a.ts]
+export class A { }
+
+//// [b.ts]
+import {A} from "./ref/a";
+export class B extends A { }
+
+
+
+//// [all.d.ts]
+declare module "ref/a" {
+    export class A {
+    }
+}
+declare module "b" {
+    import { A } from "ref/a";
+    export class B extends A {
+    }
+}

--- a/tests/baselines/reference/outModuleConcatCommonjsDeclarationOnly.symbols
+++ b/tests/baselines/reference/outModuleConcatCommonjsDeclarationOnly.symbols
@@ -1,0 +1,12 @@
+=== tests/cases/compiler/ref/a.ts ===
+export class A { }
+>A : Symbol(A, Decl(a.ts, 0, 0))
+
+=== tests/cases/compiler/b.ts ===
+import {A} from "./ref/a";
+>A : Symbol(A, Decl(b.ts, 0, 8))
+
+export class B extends A { }
+>B : Symbol(B, Decl(b.ts, 0, 26))
+>A : Symbol(A, Decl(b.ts, 0, 8))
+

--- a/tests/baselines/reference/outModuleConcatCommonjsDeclarationOnly.types
+++ b/tests/baselines/reference/outModuleConcatCommonjsDeclarationOnly.types
@@ -1,0 +1,12 @@
+=== tests/cases/compiler/ref/a.ts ===
+export class A { }
+>A : A
+
+=== tests/cases/compiler/b.ts ===
+import {A} from "./ref/a";
+>A : typeof A
+
+export class B extends A { }
+>B : B
+>A : A
+

--- a/tests/baselines/reference/outModuleConcatUnspecifiedModuleKindDeclarationOnly.js
+++ b/tests/baselines/reference/outModuleConcatUnspecifiedModuleKindDeclarationOnly.js
@@ -1,0 +1,16 @@
+//// [tests/cases/compiler/outModuleConcatUnspecifiedModuleKindDeclarationOnly.ts] ////
+
+//// [a.ts]
+export class A { } // module
+
+//// [b.ts]
+var x = 0; // global
+
+
+
+//// [out.d.ts]
+declare module "a" {
+    export class A {
+    }
+}
+declare var x: number;

--- a/tests/baselines/reference/outModuleConcatUnspecifiedModuleKindDeclarationOnly.symbols
+++ b/tests/baselines/reference/outModuleConcatUnspecifiedModuleKindDeclarationOnly.symbols
@@ -1,0 +1,8 @@
+=== tests/cases/compiler/a.ts ===
+export class A { } // module
+>A : Symbol(A, Decl(a.ts, 0, 0))
+
+=== tests/cases/compiler/b.ts ===
+var x = 0; // global
+>x : Symbol(x, Decl(b.ts, 0, 3))
+

--- a/tests/baselines/reference/outModuleConcatUnspecifiedModuleKindDeclarationOnly.types
+++ b/tests/baselines/reference/outModuleConcatUnspecifiedModuleKindDeclarationOnly.types
@@ -1,0 +1,9 @@
+=== tests/cases/compiler/a.ts ===
+export class A { } // module
+>A : A
+
+=== tests/cases/compiler/b.ts ===
+var x = 0; // global
+>x : number
+>0 : 0
+

--- a/tests/cases/compiler/outModuleConcatCommonjsDeclarationOnly.ts
+++ b/tests/cases/compiler/outModuleConcatCommonjsDeclarationOnly.ts
@@ -1,0 +1,13 @@
+// @target: ES5
+// @sourcemap: true
+// @declaration: true
+// @emitDeclarationOnly: true
+// @module: commonjs
+// @outFile: all.js
+
+// @Filename: ref/a.ts
+export class A { }
+
+// @Filename: b.ts
+import {A} from "./ref/a";
+export class B extends A { }

--- a/tests/cases/compiler/outModuleConcatUnspecifiedModuleKindDeclarationOnly.ts
+++ b/tests/cases/compiler/outModuleConcatUnspecifiedModuleKindDeclarationOnly.ts
@@ -1,0 +1,10 @@
+// @target: ES5
+// @outFile: out.js
+// @declaration: true
+// @emitDeclarationOnly: true
+
+// @Filename: a.ts
+export class A { } // module
+
+// @Filename: b.ts
+var x = 0; // global

